### PR TITLE
[SPI_HOST/DV] Regression FIX and Assertion Cov

### PR DESCRIPTION
**Pull request by visbaz**
- Regression Fixes For Over & Under Flow Test
- Assertion Coverage Test for Reg Rd

Signed-off-by: Viswanadha Bazawada <viswanadha.bazawada@ensilica.com>

### CODE REVIEW

File: hw/ip/spi_host/dv/env/seq_lib/spi_host_assert_cov_vseq.sv [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    csr_wr(.ptr(ral.intr_enable), .value(data));
+    csr_rd_check(.ptr(ral.intr_enable), .compare_value(data & 32'h3));
+    `DV_CHECK_MEMBER_RANDOMIZE_FATAL(data)
+    csr_wr(.ptr(ral.control), .value(data));
+    csr_rd_check(.ptr(ral.control), .compare_value(data & 32'he000ffff));
+    `DV_CHECK_MEMBER_RANDOMIZE_FATAL(data)
+    csr_wr(.ptr(ral.configopts[0]), .value(data));
+    csr_rd_check(.ptr(ral.configopts[0]), .compare_value(data & 32'hefffffff));
+  endtask
+endclass : spi_host_assert_cov_vseq
```
Comment (weicaiyang) [inline hw/ip/spi_host/dv/env/seq_lib/spi_host_assert_cov_vseq.sv:32]: it seems like CSR coverage collection doesn't work correctly. this may be due to the cover cfg setting. @cindychip is investigating at this.

File: hw/ip/spi_host/dv/env/seq_lib/spi_host_tx_rx_vseq.sv [hunk 2]
```diff
@@ -112,11 +112,14 @@ class spi_host_tx_rx_vseq extends spi_host_base_vseq;
...
-    end else begin
-      check_interrupts(1 << intr_state.error, 0);
+    if (intr_enable.error) begin
+      if ((enable_fld != null) && enable_fld.get()) begin
+        bit enable = `gmv(enable_fld);
+        `DV_CHECK(ral.intr_enable.predict(.value(intr_enable.error), .kind(UVM_PREDICT_DIRECT)))
+        check_interrupts(1 << intr_state.error, value & enable);
+      end else begin
+        check_interrupts(1 << intr_state.error, 0);
+      end
```
[the remark this answers is not in the document]
Comment (visbaz) [reply hw/ip/spi_host/dv/env/seq_lib/spi_host_tx_rx_vseq.sv:122]: modified

File: hw/ip/spi_host/dv/env/spi_host_scoreboard.sv [hunk 3]
```diff
@@ -420,6 +402,25 @@ class spi_host_scoreboard extends cip_base_scoreboard #(
...
+         spi_status_reg.txstall     =  get_field_val(ral.status.txstall, item.a_data);
+         spi_status_reg.tx_wm       =  get_field_val(ral.status.txwm, item.a_data);
+         spi_status_reg.rxfull      =  get_field_val(ral.status.rxfull, item.a_data);
+         spi_status_reg.rxempty     =  get_field_val(ral.status.rxempty, item.a_data);
+         spi_status_reg.rxstall     =  get_field_val(ral.status.rxstall, item.a_data);
+         spi_status_reg.byteorder   =  get_field_val(ral.status.byteorder, item.a_data);
+         spi_status_reg.rx_wm       =  get_field_val(ral.status.rxwm, item.a_data);
+         spi_status_reg.cmd_qd      =  get_field_val(ral.status.cmdqd, item.a_data);
+         spi_status_reg.rx_qd       =  get_field_val(ral.status.rxqd, item.a_data);
+         spi_status_reg.tx_qd       =  get_field_val(ral.status.txqd, item.a_data);
```
[the remark this answers is not in the document]
Comment (visbaz) [reply hw/ip/spi_host/dv/env/spi_host_scoreboard.sv:419]: modified

File: hw/ip/spi_host/dv/env/seq_lib/spi_host_assert_cov_vseq.sv (hunk 1 above)
Comment (visbaz) [reply hw/ip/spi_host/dv/env/seq_lib/spi_host_assert_cov_vseq.sv:32]: we can remove the seq once cindy finds the issue

Comment (weicaiyang) [reply hw/ip/spi_host/dv/env/seq_lib/spi_host_assert_cov_vseq.sv:32]: please remove this file. Cindy has fixed it at #12854

**Comment by visbaz** [comment]
@weicaiyang i have refactored error cmd test ..please review that.

**Comment by visbaz** [comment]
@weicaiyang or @cindychip can you help me with exclusion file. i refered to madhuri and she mentioned you created it.

**Review by weicaiyang — APPROVED** [review]
LGTM

**Comment by weicaiyang** [comment]
> @weicaiyang or @cindychip can you help me with exclusion file. i refered to madhuri and she mentioned you created it.

can you first find out why those are uncovered? Since we only require 90% for each code coverage for V2, we prefer not to apply exclusion if possible. If it's due to many unreachable coverage, we could run UNR to generate exclusion for those. But we prefer to run UNR at V3 when there is likely no more change in design.

File: hw/ip/spi_host/dv/env/seq_lib/spi_host_performance_vseq.sv [hunk 4]
```diff
@@ -33,10 +33,38 @@ class spi_host_performance_vseq extends spi_host_smoke_vseq;
...
+  endtask : body
+
   virtual task start_spi_host_trans(int num_transactions, bit wait_ready = 1'b1);
     cfg.seq_cfg.std_en  = 1;
     cfg.seq_cfg.dual_en = 1;
     cfg.seq_cfg.quad_en = 1;
     super.start_spi_host_trans(num_trans);
   endtask
+
+  virtual task generate_transaction();
```
Comment (weicaiyang) [inline hw/ip/spi_host/dv/env/seq_lib/spi_host_performance_vseq.sv:64]: I don't see we use this task in this vseq

Can you split new added vseq to another PR? They aren't related to this PR

File: hw/ip/spi_host/dv/env/seq_lib/spi_host_status_stall_vseq.sv [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    fork
+      begin : isolation_fork1
+        fork
+          start_reactive_seq();
+        join_none
+
+        begin
+          for (int i = 0; i < SPI_HOST_RX_DEPTH; i++) begin
+            check_event(ral.status.rxfull, 0);
+            if (i < SPI_HOST_RX_DEPTH - 1) check_event(ral.status.rxfull, 0);
```
Comment (weicaiyang) [inline hw/ip/spi_host/dv/env/seq_lib/spi_host_status_stall_vseq.sv:63]: we probably don't need this line as we already checked at line 62

File: hw/ip/spi_host/dv/env/seq_lib/spi_host_status_stall_vseq.sv [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      end
+      begin
+        read_rx_fifo();
+      end
+    join
+
+    gen_trans();
+    command_snd = command[0];
+    snd_cmd(command_snd);
+    csr_rd_check(.ptr(ral.status.txstall), .compare_value(1'b1));
```
Comment (weicaiyang) [inline hw/ip/spi_host/dv/env/seq_lib/spi_host_status_stall_vseq.sv:51]: perhaps code sequence like this
1. get to txfull and check txstall = 0
2. add one more item, check txstall = 1

File: hw/ip/spi_host/dv/env/seq_lib/spi_host_status_stall_vseq.sv [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+        begin
+          for (int i = 0; i < SPI_HOST_RX_DEPTH; i++) begin
+            check_event(ral.status.rxfull, 0);
+            if (i < SPI_HOST_RX_DEPTH - 1) check_event(ral.status.rxfull, 0);
+            spi_send_trans(1);
+            csr_spinwait(.ptr(ral.status.txqd), .exp_data(0));
+            csr_spinwait(.ptr(ral.status.rxqd), .exp_data(i+1));
+          end
+          check_event(ral.status.rxfull, 1);
```
Comment (weicaiyang) [inline hw/ip/spi_host/dv/env/seq_lib/spi_host_status_stall_vseq.sv:68]: also check rxstall is 0 here.

**Comment by weicaiyang** [comment]
please fix this lint error
> Whitespace check failed. Please run util/fix_trailing_whitespace.py on the above files.
